client/starwhale/api/_impl/data_store.py

```diff
@@ -386,151 +429,168 @@ def serialize(self, value: Any) -> Any:
-        raise RuntimeError(f"value should be a list: {value}")
-
-    def decode(self, value: Any) -> Any:
-        if value is None:
-            return None
-        if isinstance(value, list):
-            return [self.element_type.decode(element) for element in value]
-        raise RuntimeError(f"value should be a list: {value}")
+    def __hash__(self) -> int:
+        return hash(self.name)
 
-    def encode_type_encoded_value(self, value: Any, raw_value: bool = False) -> Any:
-        if value is None:
-            return {"type": str(self), "value": None}
-        if isinstance(value, list):
-            return {
-                "type": "LIST",
-                "value": [
-                    self.element_type.encode_type_encoded_value(element, raw_value)
-                    for element in value
-                ],
-            }
-        raise RuntimeError(f"value should be a list: {value}")
 
-    def decode_from_type_encoded_value(self, value: Any) -> Any:
-        value = value["value"]
-        if value is None:
-            return None
-        if isinstance(value, list):
-            return [
-                self.element_type.decode_from_type_encoded_value(element)
-                for element in value
-            ]
-        raise RuntimeError(f"value should be a list: {value}")
+class SwListType(SwCompositeType):
+    def __init__(
+        self,
+        element_types: List[SwType] | SwType,
+        sparse_types: Dict[int, SwType] | None = None,
+        _tuple: bool = False,
+    ) -> None:
+        if not isinstance(element_types, list):
+            element_types = [element_types]
+        if len(element_types) == 0:
+            element_types = [UNKNOWN]
+
+        self.main_type: SwType = element_types[0]
+        self.sparse_types: Dict[int, SwType] = sparse_types or {}
+
+        # find the most common type
+        types: Dict[SwType, int] = {}
+        for typ in element_types:
+            if typ in types:
+                types[typ] += 1
+            else:
+                types[typ] = 1
 
-    def __str__(self) -> str:
-        return f"[{self.element_type}]"
+        if len(types) > 1:
 
-    def __eq__(self, other: Any) -> bool:
-        if isinstance(other, SwListType):
-            return self.element_type == other.element_type
-        return False
+            def count_then_type(x: Tuple[SwType, int], y: Tuple[SwType, int]) -> int:
+                if x[1] == y[1]:
+                    return x[0] > y[0] and 1 or -1
+                return x[1] > y[1] and 1 or -1
 
+            sorted_types = sorted(
+                types.items(), reverse=True, key=cmp_to_key(count_then_type)
+            )
+            most_common_type = sorted_types[0][0]
+            self.main_type = most_common_type
+            assert isinstance(element_types, list)
+            for i, typ in enumerate(element_types):
+                if typ != most_common_type:
```

Comment:
Would this lead to this error: original list is [1,'2','3', 4.0], but when i want to get the type for index of 2(str), then i find the float type(because the main type is str, and sparse_types is [int, float])?